Hi @Dracylfrr, thanks for the contribution! I reviewed this and there's a blocking correctness bug. I built the branch and ran `IsEndOfMonthFunctionFactoryTest` locally — `testNull` fails, so this needs a fix before it can merge.

## Critical — `is_end_of_month(NULL)` returns `true` instead of `false`

`IsEndOfMonthFunctionFactory.java:67-71`:
```java
public boolean getBool(Record rec) {
    final long value = arg.getTimestamp(rec);
    return driver.getDayOfMonth(value) == driver.getDaysPerMonth(value);
}
```

There's no `Numbers.LONG_NULL` guard. For a NULL timestamp both driver methods return `Numbers.INT_NULL`:
- `MicrosTimestampDriver.getDayOfMonth(LONG_NULL)` → `INT_NULL` (`MicrosTimestampDriver.java:401-403`)
- `MicrosTimestampDriver.getDaysPerMonth(LONG_NULL)` → `INT_NULL` (`MicrosTimestampDriver.java:428-430`)
- (same in `NanosTimestampDriver` at lines 399-401 / 426-428)

So the function computes `INT_NULL == INT_NULL` → **`true`** for NULL. This is not just the literal-`null` case — **any NULL row in a real timestamp column** returns a spurious `true`, so `WHERE is_end_of_month(ts)` would wrongly select NULL rows.

Reproduced by running your own test:
```
[ERROR] IsEndOfMonthFunctionFactoryTest.testNull:36 expected:<is_end_of_month
false
> but was:<is_end_of_month
true
>
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0
```

The sibling function `IsLeapYearFunctionFactory.java:67-74` handles this correctly. Suggested fix (matches that precedent and short-circuits the NULL decomposition):
```java
import io.questdb.std.Numbers;
...
@Override
public boolean getBool(Record rec) {
    final long value = arg.getTimestamp(rec);
    if (value == Numbers.LONG_NULL) {
        return false;
    }
    return driver.getDayOfMonth(value) == driver.getDaysPerMonth(value);
}
```

The non-NULL logic (`dayOfMonth == daysPerMonth`) is correct — `testTimestamp` and `testTimestampNs` both pass.

## Moderate

**1. Redundant date decomposition (performance).** `getDayOfMonth(value)` and `getDaysPerMonth(value)` are each self-contained: both independently call `getYear` → `isLeapYear` → `getMonthOfYear` before their final step (`MicrosTimestampDriver.java:404-407` and `431-434`). So `getBool` computes the year/month/leap decomposition **twice per row**, with `getYear` (the dominant cost) running twice. On a large `WHERE` scan this doubles the per-row date-math cost. A clean fix would add a driver primitive like `boolean isLastDayOfMonth(long timestamp)` that decomposes once (keeping the micro/nano abstraction intact). Not a correctness blocker.

**2. Test coverage gaps.** All three tests pass constant literals, which the optimizer constant-folds at compile time — so the actual per-row `getBool(Record)` path against a real column is never exercised. Please add a case that reads a timestamp column from a table with month-end, mid-month, and NULL rows. Also missing: last day of a 31-day month at the year boundary (e.g. `2025-12-31`), a NULL `timestamp_ns` case, and usage inside a `WHERE` clause.

## Minor

- **Not added to `core/src/main/resources/function_list.txt`.** Functionally harmless — the scanner auto-discovers factories by classpath reflection and the list only feeds an overload-ordering map — but every other date function is listed there, so add it for consistency.
- **No trailing newline** at EOF in both new files.
- **Redundant `.ddl(null)`** on every test assertion — `ddl` already defaults to null, so the call is a no-op. Drop it.
- **PR has no labels** — add `SQL` and `New feature`.

Once the NULL guard is added the test will pass; happy to take another look after.

--- a/core/src/main/java/io/questdb/griffin/engine/functions/date/IsEndOfMonthFunctionFactory.java
+++ b/core/src/main/java/io/questdb/griffin/engine/functions/date/IsEndOfMonthFunctionFactory.java
@@ -34,6 +34,7 @@ import io.questdb.griffin.SqlExecutionContext;
 import io.questdb.griffin.engine.functions.BooleanFunction;
 import io.questdb.griffin.engine.functions.UnaryFunction;
 import io.questdb.std.IntList;
+import io.questdb.std.Numbers;
 import io.questdb.std.ObjList;
 
 public class IsEndOfMonthFunctionFactory implements FunctionFactory {
@@ -44,7 +45,13 @@ public class IsEndOfMonthFunctionFactory implements FunctionFactory {
     }
 
     @Override
-    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
+    public Function newInstance(
+            int position,
+            ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) {
         final Function arg = args.getQuick(0);
         return new Func(arg, ColumnType.getTimestampDriver(ColumnType.getTimestampType(arg.getType())));
     }
@@ -67,6 +74,9 @@ public class IsEndOfMonthFunctionFactory implements FunctionFactory {
         @Override
         public boolean getBool(Record rec) {
             final long value = arg.getTimestamp(rec);
+            if (value == Numbers.LONG_NULL) {
+                return false;
+            }
             return driver.getDayOfMonth(value) == driver.getDaysPerMonth(value);
         }
 

--- a/core/src/main/resources/function_list.txt
+++ b/core/src/main/resources/function_list.txt
@@ -275,6 +275,7 @@ io.questdb.griffin.engine.functions.date.MicrosOfMillsFunctionFactory
 io.questdb.griffin.engine.functions.date.MillisOfSecondFunctionFactory
 io.questdb.griffin.engine.functions.date.NanosOfMicrosFunctionFactory
 io.questdb.griffin.engine.functions.date.IsLeapYearFunctionFactory
+io.questdb.griffin.engine.functions.date.IsEndOfMonthFunctionFactory
 io.questdb.griffin.engine.functions.date.TimestampDiffFunctionFactory
 io.questdb.griffin.engine.functions.date.TimestampAddFunctionFactory
 io.questdb.griffin.engine.functions.date.TimestampAddWithTimezoneFunctionFactory

--- a/core/src/test/java/io/questdb/test/griffin/engine/functions/date/IsEndOfMonthFunctionFactoryTest.java
+++ b/core/src/test/java/io/questdb/test/griffin/engine/functions/date/IsEndOfMonthFunctionFactoryTest.java
@@ -28,10 +28,10 @@ import io.questdb.test.AbstractCairoTest;
 import org.junit.Test;
 
 public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
+
     @Test
     public void testNull() throws Exception {
         assertQuery("select is_end_of_month(null)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -42,7 +42,6 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
     @Test
     public void testTimestamp() throws Exception {
         assertQuery("select is_end_of_month('2024-02-29T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -50,7 +49,6 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2024-02-28T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -58,7 +56,6 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2023-02-28T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -66,7 +63,6 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2026-04-30T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -74,7 +70,6 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2026-04-29T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -82,7 +77,13 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2026-01-31T00:00:00.000000Z'::timestamp)")
-                .ddl(null)
+                .expectSize()
+                .returns("""
+                        is_end_of_month
+                        true
+                        """);
+
+        assertQuery("select is_end_of_month('2025-12-31T00:00:00.000000Z'::timestamp)")
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -91,9 +92,27 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
     }
 
     @Test
+    public void testTimestampColumn() throws Exception {
+        execute("create table x (ts timestamp)");
+        execute("insert into x values ('2024-02-29T00:00:00.000000Z')");
+        execute("insert into x values ('2024-02-28T00:00:00.000000Z')");
+        execute("insert into x values (null)");
+        execute("insert into x values ('2025-12-31T00:00:00.000000Z')");
+
+        assertQuery("select ts, is_end_of_month(ts) from x")
+                .expectSize()
+                .returns("""
+                        ts\tis_end_of_month
+                        2024-02-29T00:00:00.000000Z\ttrue
+                        2024-02-28T00:00:00.000000Z\tfalse
+                        \tfalse
+                        2025-12-31T00:00:00.000000Z\ttrue
+                        """);
+    }
+
+    @Test
     public void testTimestampNs() throws Exception {
         assertQuery("select is_end_of_month('2024-02-29T00:00:00.000000123Z'::timestamp_ns)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
@@ -101,11 +120,36 @@ public class IsEndOfMonthFunctionFactoryTest extends AbstractCairoTest {
                         """);
 
         assertQuery("select is_end_of_month('2024-02-28T00:00:00.000000123Z'::timestamp_ns)")
-                .ddl(null)
                 .expectSize()
                 .returns("""
                         is_end_of_month
                         false
+                        """);
+    }
+
+    @Test
+    public void testTimestampNsNull() throws Exception {
+        assertQuery("select is_end_of_month(null::timestamp_ns)")
+                .expectSize()
+                .returns("""
+                        is_end_of_month
+                        false
+                        """);
+    }
+
+    @Test
+    public void testTimestampColumnWhereClause() throws Exception {
+        execute("create table x (ts timestamp)");
+        execute("insert into x values ('2024-02-29T00:00:00.000000Z')");
+        execute("insert into x values ('2024-02-28T00:00:00.000000Z')");
+        execute("insert into x values (null)");
+        execute("insert into x values ('2025-12-31T00:00:00.000000Z')");
+
+        assertQuery("select ts from x where is_end_of_month(ts)")
+                .returns("""
+                        ts
+                        2024-02-29T00:00:00.000000Z
+                        2025-12-31T00:00:00.000000Z
                         """);
     }
 }

